src/canopy/llm/openai.py

```diff
@@ -164,17 +159,24 @@ def enforced_function_call(self,
 
         messages = [m.dict() for m in messages]
 
-        chat_completion = openai.ChatCompletion.create(
+        # this enforces the model to call the function
+        function_call = completion_create_params.ChatCompletionFunctionCallOptionParam(
+            name=function.name)
+
+        chat_completion = self._client.chat.completions.create(
             model=self.model_name,
             messages=messages,
-            functions=[function.dict()],
+            functions=[completion_create_params.Function(
+                name=function.name,
+                description=function.description,
+                parameters=function.parameters.dict())],
             function_call=function_call,
             max_tokens=max_tokens,
             **model_params_dict
         )
 
-        result = chat_completion.choices[0].message.function_call
-        arguments = json.loads(result["arguments"])
+        result = chat_completion.choices[0].message.function_call  # type: ignore
```

Comment:
Why did you need to `type: ignore` here?  
This seems suspicious. Are you sure we're not missing an actual required conversion (and\or something with the `response_format` arg)?

Reply:
we used cast